Assets/MixedRealityToolkit/Utilities/Rendering/MaterialInstance.cs

```diff
@@ -0,0 +1,305 @@
+                CreateInstances();
+
+                // Notify owners of the change.
+                foreach (var owner in materialOwners)
+                {
+                    (owner as IMaterialInstanceOwner)?.OnMaterialChanged(this);
+                }
+            }
+        }
+
+        private void OnDestroy()
+        {
+            if (CachedRenderer != null)
+            {
+                CachedRenderer.sharedMaterials = defaultMaterials;
+            }
+
+            DestroyMaterials(instanceMaterials);
+            instanceMaterials = null;
+        }
+
+#endregion MonoBehaviour Implementation
+
+        private void AcquireInstances()
+        {
+            if (CachedRenderer != null)
+            {
+                if (!MaterialsMatch(CachedRenderer.sharedMaterials, instanceMaterials))
+                {
+                    CreateInstances();
+                }
+            }
+        }
+
+        private void CreateInstances()
+        {
+            DestroyMaterials(instanceMaterials);
+            instanceMaterials = InstanceMaterials(defaultMaterials);
+
+            if (CachedRenderer != null)
+            {
+                CachedRenderer.sharedMaterials = instanceMaterials;
+            }
+
+            materialsInstanced = true;
+        }
+
+        private static bool MaterialsMatch(Material[] a, Material[] b)
+        {
+            if (a?.Length != b?.Length)
+            {
+                return false;
+            }
+
+            for (int i = 0; i < a?.Length; ++i)
+            {
+                if (a[i] != b[i])
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static Material[] InstanceMaterials(Material[] source)
+        {
+            if (source == null)
+            {
+                return null;
+            }
+
+            var output = new Material[source.Length];
+
+            for (var i = 0; i < source.Length; ++i)
+            {
+                if (source[i] != null)
+                {
+                    Debug.Assert(!IsInstanceMaterial(source[i]), "A material which is already instanced was instanced multiple times.");
+
+                    output[i] = new Material(source[i]);
+                    output[i].name += instancePostfix;
+                }
+            }
+
+            return output;
+        }
+
+        private static void DestroyMaterials(Material[] materials)
+        {
+            if (materials != null)
+            {
+                for (var i = 0; i < materials.Length; ++i)
+                {
+                    DestorySafe(materials[i]);
+                }
+            }
+        }
+
+        private static bool IsInstanceMaterial(Material material)
+        {
+            return ((material != null) && material.name.Contains(instancePostfix));
+        }
+
+        private static void DestorySafe(UnityEngine.Object toDestroy)
+        {
+            if (toDestroy != null)
```

Comment:
Should this maybe be like GameObjectExtensions.DestroyGameObject?

Reply:
I went to go use that but it looks to only support GameObjects? Do we need a UnityObject.Extensions?

Reply:
Yea sorry, I meant consolidating the destroy safe and GameObjectExtensions by creating an ObjectExtensions

Reply:
Ah okay, we should open an issue for this. Would be nice to delete gameObjects and components in a similar fashion.